code/messaging/src/phone/java/com/adobe/marketing/mobile/InAppNotificationHandler.java

```diff
@@ -9,98 +9,321 @@
+     * by the Optimize extension for retrieving offers
+     *
+     * @return a base64 encoded JSON string to be used by the Optimize extension for retrieving offers
+     */
+    private String getEncodedDecisionScopeForAppId() {
+        final byte[] decisionScopeBytes = String.format("{\"%s\":\"%s\"}", MessagingConstants.EventDataKeys.Optimize.XDM_NAME, offersConfig.applicationId).getBytes(StandardCharsets.UTF_8);
+
+        final EncodingService encodingService = MobileCore.getCore().eventHub.getPlatformServices().getEncodingService();
+        return new String(encodingService.base64Encode(decisionScopeBytes));
+    }
+
+    /**
+     * Converts the rules json present in the Edge response event payload into a list of rules then loads them in the {@link RulesEngine}.
+     *
+     * @param payload A {@link Map<String, Variant>} containing the personalization decision payload retrieved via the Optimize extension.
+     */
+    void handleOfferNotificationPayload(final Map<String, Variant> payload) {
+        if (MessagingUtils.isMapNullOrEmpty(payload)) {
+            Log.warning(LOG_TAG, "%s - Aborting handling of the Offers IAM payload because it is null or empty.", SELF_TAG);
             return;
         }
 
-        try {
-            final JSONObject activity = payload.getJSONObject(MessagingConstants.EventDataKeys.Offers.ACTIVITY);
-            final JSONObject placement = payload.getJSONObject(MessagingConstants.EventDataKeys.Offers.PLACEMENT);
-            final String offerActivityId = (String) activity.get(MessagingConstants.EventDataKeys.Offers.ID);
-            final String offerPlacementId = (String) placement.get(MessagingConstants.EventDataKeys.Offers.ID);
-            if (!offerActivityId.equals(POC_ACTIVITY_ID_MULTI) || !offerPlacementId.equals(POC_PLACEMENT_ID_MULTI)) {
-                Log.warning(LOG_TAG, "handleOfferNotification - ignoring Offers IAM payload, the expected offer activity id or placement id is missing. ");
+        // if we have an activity and placement id present in the manifest use the id's to validate retrieved offers
+        if (!StringUtils.isNullOrEmpty(offersConfig.activityId) && !StringUtils.isNullOrEmpty(offersConfig.placementId)) {
+            Log.trace(LOG_TAG, "%s - Activity id (%s) and placement id (%s) were found in the manifest. Using these identifiers to validate offers.", SELF_TAG, offersConfig.activityId, offersConfig.placementId);
+            final Map<String, String> activity;
+            final Map<String, String> placement;
+            final Object activityMap = payload.get(MessagingConstants.EventDataKeys.Optimize.ACTIVITY);
+            final Object placementMap = payload.get(MessagingConstants.EventDataKeys.Optimize.PLACEMENT);
+            try { // need to convert the payload map depending on the source of the offers (optimize response event or previously cached offers)
+                if (activityMap instanceof Variant) {
```

Comment:
In the case of the payload being loaded from the cache, the payload map contains a MapVariant rather than a Map<String, String>. If the payload is from an actual optimize response then I don't need to unwrap the map as it is in the expected Map<String, String> format.